python/src/iceberg/expressions/base.py

```diff
@@ -138,6 +75,24 @@ def __invert__(self) -> "BooleanExpression":
         ...
 
 
+class BoundPredicate(ABC):
+    def __init__(self, term, literal):
+        """A concrete predicate must have a `term` and `literal`"""
+        self._term = term
+        self._literal = literal
+
+
+class UnboundPredicate:
+    def __init__(self, term, literal):
+        """A concrete predicate must have a `term` and `literal`"""
+        self._term = term
+        self._literal = literal
```

Comment:
Is there value in having this `__init__` method? I don't think `In` delegates to it.

Also, every `UnboundPredicate` will have a `term`, but not all will have one or more literals. `IsNull`, `NotNull`, `IsNan`, and `NotNan` are all unary and have no term. `In` and `NotIn` have more than one literal. And the comparison predicates, like `LessThan` and `Equal` have exactly one literal. In the Java version, we model this as a list or iterable of literals and ensure requirements are handled when binding. Then the bound versions are more strict about the literals.